### ISSUE TITLE
Does shellinabox support chunked encoding?

### ISSUE THREAD

**Issue by pragatir** [shellinabox/shellinabox#321]
I am getting "Bad request' error when I set chunked-encoding in my application for POST calls.
Is there an issue with the support of chunked encoding?

**Comment by KLuka** [comment]
I don't think that this is supported on server side. Are there any benefits if we implement support for this? Could you maybe explain why this would be needed?

**Comment by pragatir** [comment]
We need it because we use REST client to connect to shellinabox which uses CHUNKED-ENCODING by default for all requests.
It would help in all those cases where large amount of data is to be sent in the request.
Also, shellinabox ultimately has a HTTP server in-built, so ideally it should support standard HTTP headers.

**Comment by pragatir** [comment]
Any thoughts? Can we have this?

**Comment by KLuka** [comment]
> We need it because we use REST client to connect to shellinabox which uses CHUNKED-ENCODING by default for all requests.

I must admit that I didn't took to much time to find what should be done in shellinabox code to support chunked encoding, but I think that this could be a lot of work... (?)

> Also, shellinabox ultimately has a HTTP server in-built, so ideally it should support standard HTTP headers.

It is true that we have a built in HTTP server, but this server is specialized for shellinabox client application. It would be great if shellinabox supported for whole HTTP protocol, but that is just to much work.

>  It would help in all those cases where large amount of data is to be sent in the request.

Currently my personal priority is implementing WebSocket support, because I think that some of current problems/limitation, even one mentioned above, will be solved with this feature. But there is just not enough free time on my hands right now... :)

So, if you or anyone else is willing to work on this issue I would be glad to examine and pull your patches. Also feel free to post more comments or information regarding this topic in this thread...

**Comment by uudruid74** [comment]
What if shellinaboxd ran under control of the web-server, as a plugin-in like the way php-fpm and such run?  This wouldn't help with Websockets (I think WebRTC could be a better approach anyway), but this may be a cleaner solution as far as installing, supporting SSL, chunked-encoding, etc, etc.  I think most people using shellinabox already run a web server anyway, right?

**Comment by a-detiste** [comment]
This means your webserver would need to run as root ?

**Comment by uudruid74** [comment]
@a-detiste  No.  Definately not.  The FastCGI protocol basically handles all the HTTP protocol, but page generation is handled by another process.   In standard CGI, the other process is reloaded for each request.  In FastCGI, the CGI is handled by a long-running process that communicates with the web server so that the a new process doesn't have to load, and in the case of scripts like perl/php, the script stays compiled without reloading and reparsing.   Basically, instead of having shellinaboxd talk HTTP to the user, it talks FastCGI to a web server.  The web server handles the HTTP protocol and SSL as an unprivaleged user and shellinaboxd would have root permission as it does now.  This means we can have the real web server serve the static files (vt100.js, beep.wav, the html files, css files, etc), won't have an internet-facing open-port, won't need to deal with SSL, etc.  It would be a simplification
